File: app/models/game/turbo_bar.rb
module Game
  # Shared resource behind turbo boost, jump jets and rockets. Draining it is what
  # makes "boost or shoot" a real choice.
  #
  # Recharge is suspended for `recharge_delay` after any draw. That delay is what
  # keeps the specified "10 rockets on a full bar" honest: rockets fire at a 100ms
  # floor, so a delay longer than the cooldown means sustained fire never earns
  # free shots mid-burst.
  class TurboBar
    # Per-frame accumulation drifts by ~1e-14 over a full recharge. Without a snap the
    # bar would sit a hair under capacity forever and never read as full.
    EPSILON = 1e-9

    attr_reader :capacity, :recharge_rate, :recharge_delay, :level

    def initialize(capacity:, recharge_rate:, recharge_delay:)
      @capacity = capacity.to_f
      @recharge_rate = recharge_rate.to_f
      @recharge_delay = recharge_delay.to_f
      @level = @capacity
      @since_draw = Float::INFINITY
    end

    def full?
      level >= capacity - EPSILON
    end

    def empty?
      level <= EPSILON
    end

    def fraction
      return 0.0 if capacity.zero?
      level / capacity
    end

    # Returns false and leaves the level untouched when the bar cannot cover the cost.
    def draw(amount)
      amount = amount.to_f
      return false if amount > level

      @level -= amount
      @since_draw = 0.0
      true
    end

    def update(dt)
      dt = dt.to_f
      @since_draw += dt

      idle = @since_draw - recharge_delay
      return level if idle <= 0

      # Only the portion of this step that fell after the delay expired earns charge.
      charged = level + recharge_rate * [ idle, dt ].min
      @level = charged >= capacity - EPSILON ? capacity : charged
    end
  end
end

File: app/javascript/game/turbo_bar.js
// Faithful port of Game::TurboBar. The bar changes every frame so its state must live
// client-side, but the semantics are Ruby's and are covered by a parity system test.
const EPSILON = 1e-9

export class TurboBar {
  constructor({ capacity, recharge_rate, recharge_delay }) {
    this.capacity = capacity
    this.rechargeRate = recharge_rate
    this.rechargeDelay = recharge_delay
    this.level = capacity
    this.sinceDraw = Infinity
  }

  get full() { return this.level >= this.capacity - EPSILON }
  get empty() { return this.level <= EPSILON }
  get fraction() { return this.capacity === 0 ? 0 : this.level / this.capacity }

  draw(amount) {
    if (amount > this.level) return false
    this.level -= amount
    this.sinceDraw = 0
    return true
  }

  update(dt) {
    this.sinceDraw += dt
    const idle = this.sinceDraw - this.rechargeDelay
    if (idle <= 0) return this.level

    const charged = this.level + this.rechargeRate * Math.min(idle, dt)
    this.level = charged >= this.capacity - EPSILON ? this.capacity : charged
    return this.level
  }
}
